examples/ann/annoy_nsmlib_example.ipynb

```diff
@@ -0,0 +1,725 @@
+      "        Scream (1996)\n",
+      "        Contact (1997)\n",
+      "User 25\n",
+      "     Known positives:\n",
+      "        Toy Story (1995)\n",
+      "        Twelve Monkeys (1995)\n",
+      "        Dead Man Walking (1995)\n",
+      "     Recommended:\n",
+      "        Independence Day (ID4) (1996)\n",
+      "        Star Wars (1977)\n",
+      "        Fargo (1996)\n",
+      "User 450\n",
+      "     Known positives:\n",
+      "        Kolya (1996)\n",
+      "        Devil's Own, The (1997)\n",
+      "        Contact (1997)\n",
+      "     Recommended:\n",
+      "        Murder at 1600 (1997)\n",
+      "        Scream (1996)\n",
+      "        Cop Land (1997)\n"
+     ]
+    }
+   ],
+   "source": [
+    "sample_recommendation_original(model, movielens, [3, 25, 450])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "So pretty similar! How about speed?"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "100 loops, best of 3: 3.85 ms per loop\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%timeit\n",
+    "sample_recommendation_original(model, movielens, [3, 25, 450], print_output=False)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "The slowest run took 5.66 times longer than the fastest. This could mean that an intermediate result is being cached.\n",
+      "100 loops, best of 3: 2.55 ms per loop\n"
+     ]
+    }
+   ],
+   "source": [
+    "%%timeit\n",
+    "sample_recommendation([3,25,450], model, movielens, print_output=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "So a small difference at thes volumes, but let's see how it compares to an sklearn distance method as n increases:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 31,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def recommend_item_for_item(item_vecotrs):\n",
+    "    distances = pairwise_distances(item_vectors[0].reshape(1, -1), Y=item_vectors)\n",
+    "    top_items = distances.argsort()[:10]\n",
+    "    return top_items\n",
+    "\n",
+    "def time_recommend_item_for_item():\n",
+    "    start = time.time()\n",
+    "    recommend_item_for_item(item_vectors)\n",
+    "    end = time.time()\n",
+    "    return end - start\n",
+    "\n",
+    "def time_Annoy_recommend_item_for_item():\n",
+    "    start = time.time()\n",
+    "    temp_t.get_nns_by_vector(item_vectors[0], 10)\n",
+    "    end = time.time()\n",
+    "    return end - start\n",
+    "\n",
+    "def time_nms_recommend_item_for_item():\n",
+    "    start = time.time()\n",
+    "    nms_idx.knnQuery(item_vectors[0])[0]\n",
```

Comment:
You should explicitly pass `nms_idx`, `item_vectors` as inputs of this function, otherwise the `nms_index` that gets used is the one from several cells before not the one defined ~20 lines below.

Maybe also generally re-factor this whole section, to have setup functions for each algorithm,  make `response_times = {'annoy': [], 'nms': []..}` use a for loop over method name, and compute the timing in this second inner loop.

Reply:
Yes much better.